File: Java/프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        int[][] visited = new int[n + 1][m + 1];

        for (int[] puddle : puddles) {
            visited[puddle[1]][puddle[0]] = -1;
        }
        visited[1][1] = 1;

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (visited[i][j] == -1) {
                    continue;
                }
                if (visited[i - 1][j] != -1) {
                    visited[i][j] += visited[i - 1][j];
                    visited[i][j] %= 1000000007;
                }
                if (visited[i][j - 1] != -1) {
                    visited[i][j] += visited[i][j - 1];
                    visited[i][j] %= 1000000007;
                }
            }
        }
        answer = visited[n][m];

        return answer;
    }
}

File: Python/프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    route = [[0] * (m + 1) for _ in range(n + 1)]

    for i, j in puddles:
        route[j][i] = -1
    route[1][1] = 1

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if route[i][j] == -1:
                continue

            if route[i - 1][j] != -1:
                route[i][j] += route[i - 1][j]

            if route[i][j - 1] != -1:
                route[i][j] += route[i][j - 1]

            route[i][j] %= 1000000007

    return route[n][m]